src/result_object/index.js

```diff
@@ -56,6 +55,32 @@ class ResultObject {
         return this;
     }
 
+    /**
+     * Bumps field's warning counts and adds warning string
+     * @param {string} fieldName - The name of the field.
+     * @param {string} statement - The error string to add to the object.
+     */
+    bumpTestWarning(fieldName: string, statement: string) {
+        this.hasValidationWarnings = true;
+        this.validationWarnings[fieldName] = this.validationWarnings[fieldName] || [];
+        this.validationWarnings[fieldName].push(statement);
+        this.warnCount++;
+        this.testsPerformed[fieldName].warnCount++;
+    }
+
+    /**
+     * Bumps field's error counts and adds error string
+     * @param {string} fieldName - The name of the field.
+     * @param {string} statement - The error string to add to the object.
+     */
+    bumpTestError(fieldName: string, statement: string) {
+        this.hasValidationErrors = true;
+        this.validationErrors[fieldName] = this.validationErrors[fieldName] || [];
+        this.validationErrors[fieldName].push(statement);
+        this.failCount++;
+        this.testsPerformed[fieldName].failCount++;
+    }
```

Comment:
I am not sur about all those assignment: 
. `failCount` can be the result of a function
. `testsPerformed[fieldName].failCount` same
. `hasValidationErrors` same

This will remove from code a lot of assigment, what do you think? If you are afraid about perf, you can maybe add a benchmark

Reply:
If all this properties are equal to `f(validationErrors[fieldName])` that's mean you just need to unit test those functions and not have to test the entire object

Reply:
1. Regarding the dynamic enumeration of the counters. Really cool idea, outside the scope of this change. I don't worry about the performance. The only issue I can think of is that there will be much more logic to maintain and test. This one is as simple as it gets: `++`.

2. I prefer testing the whole Object and not just single property. I want to know that everything works together and I don't get any weird side effects.